contracts/UnionZap.sol

```diff
@@ -319,64 +363,213 @@ contract UnionZap is Ownable, UnionBase {
                 } else {
                     _swapToETH(_token, _balance, routers[_choice]);
                 }
+                routerChoices = routerChoices >> 3;
             }
-            routerChoices = routerChoices >> 3;
         }
 
-        uint256 _ethBalance = address(this).balance;
+        // slippage check
+        assert(address(this).balance > minAmountOut);
 
-        // if locking, we apply minAmount to CRV - otherwise will do on cvxCRV
-        uint256 minCrvOut = lock ? minAmountOut : 0;
-        // swap from ETH to CRV
-        uint256 _swappedCrv = _swapEthToCrv(_ethBalance, minCrvOut);
+        (bool success, ) = (tx.origin).call{value: gasRefund}("");
+        require(success, "ETH transfer failed");
+    }
 
-        uint256 _crvBalance = IERC20(CRV_TOKEN).balanceOf(address(this));
+    /// @notice Internal function used to sell output tokens for ETH
+    /// @param _token - the token to sell
+    /// @param _amount - how much of that token to sell
+    function _sell(address _token, uint256 _amount) internal {
+        if (_token == CRV_TOKEN) {
+            _crvToEth(_amount, 0);
+        } else if (_token == CVX_TOKEN) {
+            _swapToETHCurve(_token, _amount);
+        } else {
+            IERC20(_token).safeTransfer(tokenInfo[_token].swapper, _amount);
+            ISwapper(tokenInfo[_token].swapper).sell(_amount);
+        }
+    }
+
+    /// @notice Internal function used to buy output tokens from ETH
+    /// @param _token - the token to sell
+    /// @param _amount - how much of that token to sell
+    function _buy(address _token, uint256 _amount) internal {
+        if (_token == CRV_TOKEN) {
+            _ethToCrv(_amount, 0);
+        } else if (_token == CVX_TOKEN) {
+            _ethToCvx(_amount, 0);
+        } else {
+            (bool success, ) = tokenInfo[_token].swapper.call{value: _amount}(
+                ""
+            );
+            require(success, "ETH transfer failed");
+            ISwapper(tokenInfo[_token].swapper).buy(_amount);
+        }
+    }
 
+    /// @notice Swap or lock all CRV for cvxCRV
+    /// @param _minAmountOut - the min amount of cvxCRV expected
+    /// @param _lock - whether to lock or swap
+    /// @return the amount of cvxCrv obtained
+    function _toCvxCrv(uint256 _minAmountOut, bool _lock)
+        internal
+        returns (uint256)
+    {
+        uint256 _crvBalance = IERC20(CRV_TOKEN).balanceOf(address(this));
         // swap on Curve if there is a premium for doing so
-        if (!lock) {
-            _swapCrvToCvxCrv(_crvBalance, address(this), minAmountOut);
+        if (!_lock) {
+            return _swapCrvToCvxCrv(_crvBalance, address(this), _minAmountOut);
         }
         // otherwise deposit & lock
         else {
```

Comment:
Can remove `else` since we are returning in `if`

Reply:
Fixed in 927d350423444a39dba0e73631b8a6cf6d9284c9